openedx/core/djangoapps/content/learning_sequences/api/tests/test_visibility.py

```diff
@@ -0,0 +1,59 @@
+from datetime import datetime, timedelta, timezone
+
+import ddt
+import mock
+from django.contrib.auth.models import AnonymousUser, User
+from openedx.core.djangoapps.content.learning_sequences.api import get_user_course_outline, replace_course_outline
+from openedx.core.djangoapps.waffle_utils.testutils import override_waffle_flag
+from openedx.core.djangolib.testing.utils import CacheIsolationTestCase
+from openedx.features.course_experience import COURSE_ENABLE_UNENROLLED_ACCESS_FLAG
+from xmodule.course_module import COURSE_VISIBILITY_PRIVATE, COURSE_VISIBILITY_PUBLIC, COURSE_VISIBILITY_PUBLIC_OUTLINE
+from xmodule.modulestore import ModuleStoreEnum
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore.tests.django_utils import TEST_DATA_MIXED_MODULESTORE, ModuleStoreTestCase
+from xmodule.modulestore.tests.factories import CourseFactory
+
+from ..data import CourseOutlineData
+from .test_data import generate_sections
+
+
+class CourseOutlineVisibilityTestCase(ModuleStoreTestCase, CacheIsolationTestCase):
+    MODULESTORE = TEST_DATA_MIXED_MODULESTORE
+
+    def setUp(self):
+        super(CourseOutlineVisibilityTestCase, self).setUp()
+
+        course_data = {
+            "start": datetime.now() - timedelta(days=30),
+            "end": datetime.now() + timedelta(days=30),
+            "run": "2020_Summer",
+            "number": "CS101",
+            "org": "ABKUni",
+            "self_paced": True,
+            "course_visibility": "private"
+        }
+
+        self.user = User.objects.create(username="test_user")
+
+        self.course = CourseFactory.create(default_store=ModuleStoreEnum.Type.split, **course_data)
+        self.course.save()
+
+        course_outline = CourseOutlineData(
+            course_key=self.course.id,
+            title="Some Title Four Course Outline",
+            published_at=datetime(2020, 5, 20, tzinfo=timezone.utc),
+            published_version="5ebece4b69dd593d82fe2015",
+            sections=generate_sections(self.course.id, [2, 2]),
+        )
+        replace_course_outline(course_outline)
+
+    def testCourse(self):
+        # NOTE: This test function doesn't work correctly.
+        for visibility in [COURSE_VISIBILITY_PRIVATE, COURSE_VISIBILITY_PUBLIC, COURSE_VISIBILITY_PUBLIC_OUTLINE]:
+            with self.subTest(visibility=visibility):
+                with mock.patch('openedx.core.djangoapps.content.learning_sequences.api.outlines.get_course_outline') as patched_get_user_course_outline:
```

Comment:
Thinking on this a bit more, I think that we can get a simpler testing story that doesn't involve mocking with a bit of refactoring of `update_from_modulestore`, basically by separating out the 99% of that function that extracts a `CourseOutlineData` from the modulestore from the line that does the `replace_course_outline`.

So it'd look like:

```python
def get_outline_from_modulestore(course_key):
    # most of what's in update_from_modulestore now

def update_from_modulestore(course_key):
    course_outline_data = get_outline_from_modulestore(course_key)
    replace_course_outline(course_outline_data)
```

So then the only thing we test on the modulestore side is `get_outline_from_modulestore`, to make sure that it spits out the values we expect from the CourseFactory that's set up here.

Then we rely on the `test_outlines.py` style tests (where we create the CourseOutlineData we want and flip bits here and there to test behaviors) for all the permutations around waffle flag behavior and any other edge cases we want to try. So we never test end-to-end really, but we're instead testing each half separately.

Does that make sense?